Allow Project Find Results to be copied to clipboard

liuderchi opened the pull request:
Related issue [#416](https://github.com/atom/find-and-replace/issues/416).
copy result of _Project-Find_ to clipboard.

changes:
1. add `Util.copySearchResultFromPane()` to copy text from search result pane
2. add context menu item supporting right-click on project-find result page. 
   (of course can use it via command-palette)

demo:
![demo](https://cloud.githubusercontent.com/assets/4994705/17399255/9907ebb8-5a73-11e6-8ef5-2522cb7d3ba2.png)

NOTE: hidden entries under collapsed item are not copied to clipboard so that result is consistent with view.

50Wliu commented on lib/project-find-view.coffee:
```diff
@@ -183,6 +186,10 @@ class ProjectFindView extends View
 
+  copySearchResultFromPane: ->
+    atom.clipboard.write(Util.parseSearchResult())
+    atom.notifications.addInfo('Search results are copied to clipboard')
```
`Search results have been copied to clipboard`

50Wliu commented on lib/project/util.coffee:
```diff
@@ -25,3 +26,20 @@ module.exports =
     else
       "No #{if replacedPathCount? then 'more' else ''} results found for '#{@sanitizePattern(findPattern)}'"
+
+  parseSearchResult: () ->
```
`parseSearchResult: ->`

50Wliu commented on lib/project/util.coffee:
```diff
@@ -25,3 +26,20 @@ module.exports =
+    summary = $('span.preview-count', 'div.preview-pane').text()
+    searchResult.push summary, ''
+
+    $('ol.results-view.list-tree>li.path').each () ->
```
`$('ol.results-view.list-tree>li.path').each ->`

50Wliu commented on lib/project/util.coffee:
```diff
@@ -25,3 +26,20 @@ module.exports =
+    searchResult.push summary, ''
+
+    $('ol.results-view.list-tree>li.path').each () ->
+      path = $('span.path-name', @).text()
```
Replace `@` with `this` wherever it's not followed by a variable

50Wliu commented on lib/project/util.coffee:
```diff
@@ -25,3 +26,20 @@ module.exports =
+      matches = $('span.path-match-number', @).text()
+      searchResult.push path + ' ' + matches
+
+      $('li.search-result', @).filter(':visible').each () ->
```
Same thing as on line 35.

50Wliu commented:
Very nice!  Left you a few comments to resolve the lint errors.

50Wliu commented on lib/project-find-view.coffee:
```diff
@@ -146,6 +146,9 @@ class ProjectFindView extends View
       'project-find:replace-all': => @replaceAll()
 
+    @subscriptions.add atom.commands.add 'div.preview-pane',
+      'project-find:copy-search-result': @copySearchResultFromPane
```
`-results`

liuderchi replied:
I will update this in `menus/find-and-replace.cson` also.

liuderchi commented:
@50Wliu is there anything else I need to change? :smiley:

50Wliu commented:
/cc @maxbrunsfeld

jerone commented on lib/project/util.coffee:
```diff
@@ -25,3 +26,20 @@ module.exports =
+        preview = $('span.preview', this).text()
+        searchResult.push '\t' + lineNumber + '\t' + preview
+      searchResult.push ''
+    searchResult.join('\n')
```
Above part could be rewritten with plain DOM API to remove the `space-pen` dependency with minimal effort. Might help with getting this merged.

liuderchi replied:
Good Idea and I'm working on it.

liuderchi commented:
@50Wliu how do think about the format of copied content?

liuderchi commented:
@50Wliu @jerone this PR now is
  - using DOM API without `space-pen` dependency
  - having corresponding spec

Anything else I can do?

50Wliu commented:
I don't feel qualified to review this PR, but I'll try to get someone else to take a look at it.

dezman commented:
Who has context top review this,? Would love this feature.

liuderchi committed:
:art: copy search result of project-find to clipboard

- add parseSearchResult() in util.coffee with DOM api
- add copySearchResultFromPane() in project-find-view.js
- add spec in project-find-view-spec.js
- add context-menu item

liuderchi commented:
currently project-find results page shows *only visible items* in DOM tree.
so this PR cannot parse all search results in this way.

steveoh commented:
This would be pretty cool to have.

liuderchi commented:
@steveoh too bad this PR does not fit the view in current version.

steveoh commented:
this should be triaged at some point.
